File: frontend/src/types/report.ts
// --- Enum types ---

export type ReportType =
  | "saldi_conti"
  | "movimenti"
  | "entrate_uscite"
  | "riconciliazione"
  | "scadenzario"
  | "aging"
  | "cash_flow"
  | "pagamenti"
  | "controparte"
  | "personalizzato";

export type ReportFormat = "pdf" | "xlsx" | "csv";

export type ReportExecutionStatus = "in_coda" | "in_generazione" | "completato" | "errore";

export type ReportFrequency = "giornaliero" | "settimanale" | "mensile" | "trimestrale";

// --- Label maps ---

export const REPORT_TYPE_LABELS: Record<ReportType, string> = {
  saldi_conti: "Situazione Saldi",
  movimenti: "Estratto Conto",
  entrate_uscite: "Entrate e Uscite",
  riconciliazione: "Riconciliazione",
  scadenzario: "Scadenzario",
  aging: "Aging Crediti/Debiti",
  cash_flow: "Cash Flow",
  pagamenti: "Pagamenti",
  controparte: "Scheda Controparte",
  personalizzato: "Personalizzato",
};

export const REPORT_FORMAT_LABELS: Record<ReportFormat, string> = {
  pdf: "PDF",
  xlsx: "Excel",
  csv: "CSV",
};

export const REPORT_EXECUTION_STATUS_LABELS: Record<ReportExecutionStatus, string> = {
  in_coda: "In coda",
  in_generazione: "In generazione",
  completato: "Completato",
  errore: "Errore",
};

export const REPORT_FREQUENCY_LABELS: Record<ReportFrequency, string> = {
  giornaliero: "Giornaliero",
  settimanale: "Settimanale",
  mensile: "Mensile",
  trimestrale: "Trimestrale",
};

// --- Status colors ---

export const REPORT_STATUS_COLORS: Record<ReportExecutionStatus, string> = {
  in_coda: "bg-[#f2f2f2] text-[#817f7d]",
  in_generazione: "bg-[#c0c8ff] text-[#1a1a1a]",
  completato: "bg-[#8cddba] text-[#1a1a1a]",
  errore: "bg-[#fddcdc] text-[#ea0b0b]",
};

// --- Icons per tipo ---

export const REPORT_TYPE_ICONS: Record<ReportType, string> = {
  saldi_conti: "Wallet",
  movimenti: "List",
  entrate_uscite: "ArrowUpDown",
  riconciliazione: "CheckCircle",
  scadenzario: "Calendar",
  aging: "Clock",
  cash_flow: "TrendingUp",
  pagamenti: "CreditCard",
  controparte: "Users",
  personalizzato: "FileText",
};

// --- Interfaces ---

export interface ReportDefinition {
  id: string;
  company_id: string | null;
  nome: string;
  descrizione: string | null;
  tipo: ReportType;
  formato_default: ReportFormat;
  parametri_default: Record<string, unknown> | null;
  layout: Record<string, unknown> | null;
  is_system: boolean;
  is_preferito: boolean;
  sort_order: number;
  created_at: string;
  updated_at: string;
}

export interface ReportExecution {
  id: string;
  company_id: string;
  report_definition_id: string | null;
  generato_da_user_id: string | null;
  nome: string;
  tipo: string;
  formato: ReportFormat;
  stato: ReportExecutionStatus;
  parametri: Record<string, unknown> | null;
  risultato: Record<string, unknown> | null;
  file_path: string | null;
  file_size: number | null;
  errore: string | null;
  righe_totali: number | null;
  data_generazione: string;
  data_completamento: string | null;
  data_scadenza_file: string | null;
  created_at: string;
  updated_at: string;
}

export interface ReportSchedule {
  id: string;
  company_id: string;
  report_definition_id: string;
  creato_da_user_id: string | null;
  nome: string;
  frequenza: ReportFrequency;
  formato: ReportFormat;
  parametri: Record<string, unknown> | null;
  attivo: boolean;
  prossima_esecuzione: string | null;
  ultima_esecuzione: string | null;
  created_at: string;
  updated_at: string;
}

export interface ReportDefinitionListResponse {
  items: ReportDefinition[];
  total: number;
}

export interface ReportExecutionListResponse {
  items: ReportExecution[];
  total: number;
  page: number;
  page_size: number;
}

export interface ReportScheduleListResponse {
  items: ReportSchedule[];
  total: number;
}

File: backend/app/schemas/report.py
import uuid
from datetime import date, datetime
from decimal import Decimal

from pydantic import BaseModel, Field


# --- Response schemas ---

class ReportDefinitionResponse(BaseModel):
    id: uuid.UUID
    company_id: uuid.UUID | None = None
    nome: str
    descrizione: str | None = None
    tipo: str
    formato_default: str
    parametri_default: dict | None = None
    layout: dict | None = None
    is_system: bool
    is_preferito: bool
    sort_order: int
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class ReportDefinitionListResponse(BaseModel):
    items: list[ReportDefinitionResponse]
    total: int


class ReportExecutionResponse(BaseModel):
    id: uuid.UUID
    company_id: uuid.UUID
    report_definition_id: uuid.UUID | None = None
    generato_da_user_id: uuid.UUID | None = None
    nome: str
    tipo: str
    formato: str
    stato: str
    parametri: dict | None = None
    risultato: dict | None = None
    file_path: str | None = None
    file_size: int | None = None
    errore: str | None = None
    righe_totali: int | None = None
    data_generazione: datetime
    data_completamento: datetime | None = None
    data_scadenza_file: datetime | None = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class ReportExecutionListResponse(BaseModel):
    items: list[ReportExecutionResponse]
    total: int
    page: int
    page_size: int


class ReportScheduleResponse(BaseModel):
    id: uuid.UUID
    company_id: uuid.UUID
    report_definition_id: uuid.UUID
    creato_da_user_id: uuid.UUID | None = None
    nome: str
    frequenza: str
    formato: str
    parametri: dict | None = None
    attivo: bool
    prossima_esecuzione: datetime | None = None
    ultima_esecuzione: datetime | None = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class ReportScheduleListResponse(BaseModel):
    items: list[ReportScheduleResponse]
    total: int


# --- Request schemas ---

class ReportDefinitionCreateRequest(BaseModel):
    nome: str = Field(min_length=1, max_length=255)
    descrizione: str | None = None
    tipo: str = "personalizzato"
    formato_default: str = "pdf"
    parametri_default: dict | None = None
    layout: dict | None = None


class ReportDefinitionUpdateRequest(BaseModel):
    nome: str | None = Field(None, min_length=1, max_length=255)
    descrizione: str | None = None
    formato_default: str | None = None
    parametri_default: dict | None = None
    layout: dict | None = None


class ReportGenerateRequest(BaseModel):
    formato: str = "pdf"
    parametri: dict | None = None


class ReportGenerateQuickRequest(BaseModel):
    tipo: str
    formato: str = "pdf"
    parametri: dict | None = None
    nome: str | None = None


class ReportScheduleCreateRequest(BaseModel):
    report_definition_id: uuid.UUID
    nome: str = Field(min_length=1, max_length=255)
    frequenza: str
    formato: str = "pdf"
    parametri: dict | None = None


class ReportScheduleUpdateRequest(BaseModel):
    nome: str | None = Field(None, min_length=1, max_length=255)
    frequenza: str | None = None
    formato: str | None = None
    parametri: dict | None = None
    attivo: bool | None = None
